src/world.rs

```diff
@@ -569,33 +583,34 @@ impl World {
 
         unsafe {
             // Drop the components we're overwriting
-            let source_arch = &mut self.archetypes.archetypes[loc.archetype as usize];
+            let source_arch = &mut self.archetypes.archetypes[old_archetype as usize];
```

Comment:
It's confusing that `source_arch` is not derived from `source_archetype`. Maybe rename `source_archetype`, perhaps to `archetype_graph_origin` or something similarly explicit?